src/core/createCookie.js

```diff
@@ -28,35 +29,45 @@ const safeJSONParse = (object, cookieName) => {
  * read and deserialize it every time a piece of it is accessed.
  */
 const createCookieProxy = (propertyId, cookieDomain = "") => {
-  const cookieName = `${ALLOY_COOKIE_NAME}_${propertyId}`;
+  // Because the cookie proxy caches the deserialized cookie, we want to make
+  // sure there's only one cookie proxy created per cookie because multiple
+  // caches per cookies would result in caches getting stale.
+  let cookieProxy = cookieProxyByPropertyId[propertyId];
 
-  let deserializedCookie;
-  let cookieHasBeenRead = false;
+  if (!cookieProxy) {
+    const cookieName = `${ALLOY_COOKIE_NAME}_${propertyId}`;
+    let deserializedCookie;
+    let cookieHasBeenRead = false;
 
-  return {
-    get() {
-      // We don't read the cookie right when the cookie proxy is created
-      // because we don't know if the user has opted in. If the user
-      // hasn't opted in, we're legally obligated to not read the cookie.
-      // If a component tries to read something off the cookie though,
-      // we assume that the component has received word that the user
-      // has opted-in. The responsibility is on the component.
-      if (!cookieHasBeenRead) {
-        const serializedCookie = cookie.get(cookieName);
-        deserializedCookie =
-          serializedCookie && safeJSONParse(serializedCookie, cookieName);
-        cookieHasBeenRead = true;
+    cookieProxy = {
+      get() {
+        // We don't read the cookie right when the cookie proxy is created
+        // because we don't know if the user has opted in. If the user
+        // hasn't opted in, we're legally obligated to not read the cookie.
+        // If a component tries to read something off the cookie though,
+        // we assume that the component has received word that the user
+        // has opted-in. The responsibility is on the component.
```

Comment:
May be we can invert the condition, here as well.

Reply:
Sure. I added a commit with the change.